### PR TITLE
Added required member listings

### PR DESCRIPTION
**Pull request by MarkMichaelis**

### CODE REVIEW

File: src/Chapter06/Listing06.28.DefaultConstructorNoLongerAvailable.cs [hunk 1]
```diff
@@ -12,7 +12,7 @@ public static void Main()
         {
             Employee employee;
 
-            asdfasdfasdfdsaf
+             //asdfasdfasdfdsaf
```
Comment (Keboo) [inline src/Chapter06/Listing06.28.DefaultConstructorNoLongerAvailable.cs]: I had commented this line just to trigger the single error about the ctor missing parameter. Should this actually be removed?

File: src/Chapter06.Tests/CompilerAssert2.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Testing;
+using Microsoft.CodeAnalysis.Testing.Verifiers;
+using System.Collections.Immutable;
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Shared;
+
+public static class CompilerAssert2
```
Comment (Keboo) [inline src/Chapter06.Tests/CompilerAssert2.cs]: I assume at some point we should come back in and clean this up as I did it more as just an example code.

File: src/Chapter06/Listing06.35A.RequiredProperties.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
+#if NET7_0_OR_GREATER
+using System.Diagnostics.CodeAnalysis;
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter06.Listing06_35A
```
Comment (Keboo) [inline src/Chapter06/Listing06.35A.RequiredProperties.cs:3]: File scoped namespace here?

File: src/Chapter06/Listing06.35A.RequiredProperties.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        public static void Main()
+        {
+            #region EXCLUDE
+            #pragma warning disable IDE0059 // Unnecessary assignment of a value
+            #endregion EXCLUDE
+            Book book = new()
+            {
+                Isbn= "978-0135972267",
+                Title= "Harold and the Purple Crayon"
```
Comment (Keboo) [inline src/Chapter06/Listing06.35A.RequiredProperties.cs]: nit:
```suggestion
                Isbn = "978-0135972267",
                Title = "Harold and the Purple Crayon"
```

File: src/Chapter06/Listing06.35A.RequiredProperties.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        #endregion HIGHLIGHT
+        {
+            get
+            {
+                return _Title!;
+            }
+            set
+            {
+                ArgumentException.ThrowIfNullOrEmpty(
+                    value = value?.Trim()!);
```
Comment (Keboo) [inline src/Chapter06/Listing06.35A.RequiredProperties.cs]: Is it pertinent to the example that we also check for whitespace? It ends up making the code a lot more complicated than it needs to be.

File: src/Chapter06/Listing06.35B.SpecifyingRequiredMembersWithinTheObjectInitializer.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+#if NET7_0_OR_GREATER
+
+ #if INCLUDE
+using AddisonWesley.Michaelis.EssentialCSharp.Chapter06.Listing06_35A;
+ #endif // INCLUDE
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter06.Listing06_35B
```
Comment (Keboo) [inline src/Chapter06/Listing06.35B.SpecifyingRequiredMembersWithinTheObjectInitializer.cs:7]: File scope namespace

File: src/Chapter06/Listing06.35B.SpecifyingRequiredMembersWithinTheObjectInitializer.cs [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+{
+    public class Program
+    {
+        public static void Main()
+        {
+            #if INCLUDE
```
Comment (Keboo) [inline src/Chapter06/Listing06.35B.SpecifyingRequiredMembersWithinTheObjectInitializer.cs]: Do we really want to use INCLUDE for the symbol, it seems a little too close to our `#region INCLUDE`?

File: src/Chapter06/Listing06.35C.DisablingRequiredObjectInitialization.cs [hunk 8]
```diff
@@ -0,0 +1,69 @@
+#if NET7_0_OR_GREATER
+using System.Diagnostics.CodeAnalysis;
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter06.Listing06_34C
```
Comment (Keboo) [inline src/Chapter06/Listing06.35C.DisablingRequiredObjectInitialization.cs:4]: File scoped namespace?

File: src/Chapter06/Listing06.35C.DisablingRequiredObjectInitialization.cs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+{
+    public class Program
+    {
+        public static void Main()
+        {
+            Book book = new(42) { 
+                Subtitle = "The Comprehensive, Expert Guide " +
+                    "to C# for Programmers at All Levels" };
```
Comment (Keboo) [inline src/Chapter06/Listing06.35C.DisablingRequiredObjectInitialization.cs]: Any reason why these are capitalized?
```suggestion
                    "to C# for programmers at all levels" };
```

File: src/Shared/CompilerAssert.cs [hunk 10]
```diff
@@ -176,23 +176,24 @@ public static async Task Compile2Async(string[] fileNames, string[] errorIds, st
                     (string result, string item) =>
                         result += $"\n{item}");
 
-            return await ExpectErrorsAsync(sourceCode, diagnostics);
+            return ExpectErrorsAsync(sourceCode, diagnostics);
         }
 
-        async static public Task<CompileError[]> ExpectErrorsAsync(
+        public static CompileError[] ExpectErrorsAsync(
```
Comment (Keboo) [inline src/Shared/CompilerAssert.cs]:
```suggestion
        public static CompileError[] ExpectErrors(
```

**Review by Keboo — APPROVED** [review]

File: src/Chapter06/Listing06.35C.DisablingRequiredObjectInitialization.cs (hunk 8 above)
Comment (MarkMichaelis) [reply src/Chapter06/Listing06.35C.DisablingRequiredObjectInitialization.cs:4]: I'll make all these changes in one swell foop at the end.

File: src/Chapter06/Listing06.35B.SpecifyingRequiredMembersWithinTheObjectInitializer.cs (hunk 7 above)
Comment (MarkMichaelis) [reply src/Chapter06/Listing06.35B.SpecifyingRequiredMembersWithinTheObjectInitializer.cs]: @Keboo: How about `#if COMPILEERROR`?

Comment (Keboo) [reply src/Chapter06/Listing06.35B.SpecifyingRequiredMembersWithinTheObjectInitializer.cs]: That works for me

File: src/Chapter06.Tests/Listing06.28.Tests.cs [hunk 11]
```diff
@@ -1,18 +1,17 @@
...
-namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter06.Listing06_28.Tests
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter06.Listing06_28.Tests;
+
+[TestClass]
+public class ProgramTests
 {
-    [TestClass]
-    public class ProgramTests
+    [TestMethod]
+    [Ignore]
```
Comment (Keboo) [inline src/Chapter06.Tests/Listing06.28.Tests.cs:10]: Can we now drop the `[Ignore]`here?

File: src/Chapter06.Tests/Listing06.35B.Tests.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
+﻿using AddisonWesley.Michaelis.EssentialCSharp.Shared.Tests;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter06.Listing06_35B.Tests;
+
+[TestClass]
+public class ProgramTests
+{
+    [TestMethod]
+    [Ignore]
```
Comment (Keboo) [inline src/Chapter06.Tests/Listing06.35B.Tests.cs:10]: Does this still need to be ignored?

File: src/Chapter06.Tests/Listing06.36.UsingNotNullWhenAndNotNullIfNotNullAttributes.Tests.cs [hunk 13]
```diff
@@ -1,34 +1,33 @@
...
-        public void TryGetAsText_Given4_ReturnFour()
+        if(NullabilityAttributesExamined.TryGetDigitAsText('4', out string? text))
         {
-            if(NullabilityAttributesExamined.TryGetDigitAsText('4', out string? text))
-            {
-                Assert.AreEqual<string>("four", text);
-            }
+            Assert.AreEqual<string>("four", text);
         }
+    }
```
Comment (Keboo) [inline src/Chapter06.Tests/Listing06.36.UsingNotNullWhenAndNotNullIfNotNullAttributes.Tests.cs]: You shouldn't have conditions in the unit test. Better to Assert that `TryGetDigitAsText` returns true, and then assert that `text` equals `"four"`

File: src/Chapter06.Tests/Listing06.37.UsingTheMaybeNullAttribute.Tests.cs [hunk 14]
```diff
@@ -1,28 +1,27 @@
...
+    public void GetObject_CollectionWithItem_ReturnItem()
     {
-        [TestMethod]
-        public void GetObject_CollectionWithItem_ReturnItem()
-        {
-            NullabilityAttributesExamined.GetObject(
-                new string[] { "item" }, item => item is object);
-        }
+        NullabilityAttributesExamined.GetObject(
+            new string[] { "item" }, item => item is object);
```
Comment (Keboo) [inline src/Chapter06.Tests/Listing06.37.UsingTheMaybeNullAttribute.Tests.cs]: Assert-less test should at least have a comment. Should this be check the return is not null?

File: src/Chapter06/Listing06.42.DefiningAStaticMethodOnDirectoryInfo.cs [hunk 15]
```diff
@@ -3,71 +3,72 @@ namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter06.Listing06_42
...
-                if(!Directory.Exists(target))
-                {
-                    Directory.CreateDirectory(target);
-                }
+                target += Path.DirectorySeparatorChar;
+            }
+            if (!Directory.Exists(target))
+            {
+                Directory.CreateDirectory(target);
+            }
```
Comment (Keboo) [inline src/Chapter06/Listing06.42.DefiningAStaticMethodOnDirectoryInfo.cs]:
```suggestion
            Directory.CreateDirectory(target);
```

File: src/Chapter06/Listing06.42.DefiningAStaticMethodOnDirectoryInfo.cs [hunk 16]
```diff
@@ -3,71 +3,72 @@ namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter06.Listing06_42
...
 
-                for(int i = 0; i < searchPattern.Length; i++)
+            for (int i = 0; i < searchPattern.Length; i++)
+            {
+                foreach (string file in
+                    Directory.GetFiles(
```
Comment (Keboo) [inline src/Chapter06/Listing06.42.DefiningAStaticMethodOnDirectoryInfo.cs]: Can we use `Directory.EnumerateFiles` here?

File: src/Chapter06/Listing06.42.DefiningAStaticMethodOnDirectoryInfo.cs [hunk 17]
```diff
@@ -3,71 +3,72 @@ namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter06.Listing06_42
...
                 }
+            }
 
-                // Copy subdirectories (recursively)
-                if(option == SearchOption.AllDirectories)
+            // Copy subdirectories (recursively)
+            if (option == SearchOption.AllDirectories)
+            {
+                foreach (string element in
+                    Directory.GetDirectories(
```
Comment (Keboo) [inline src/Chapter06/Listing06.42.DefiningAStaticMethodOnDirectoryInfo.cs]: Can we use `Directory.EnumerateDirectories` here?

**Review by Keboo — APPROVED** [review]

File: src/Chapter06.Tests/Listing06.28.Tests.cs (hunk 11 above)
Comment (MarkMichaelis) [reply src/Chapter06.Tests/Listing06.28.Tests.cs:10]: No... these are still not passing and I didn't investigate.

File: src/Chapter06.Tests/Listing06.35B.Tests.cs (hunk 12 above)
Comment (MarkMichaelis) [reply src/Chapter06.Tests/Listing06.35B.Tests.cs:10]: These are still not passing and I didn't investigate.

**Pull request merged**